### PR TITLE
api-put-object-streaming: Removed unnecessary `partBuf` copy to save allocations; no race detected.

### PR DESCRIPTION
**Pull request by bwplotka**
Hi and thanks for the great work so far. 👋🏽 

Following our discussion in https://github.com/minio/mc/issues/3376#issuecomment-806013131 I performed extended testing, benchmarking and profiling to figure out what's best for `minio-go` and the Thanos [issue](https://github.com/thanos-io/thanos/issues/3967). 

Proving tests:
* Race [test](https://github.com/thanos-io/thanos/pull/5474/files#diff-5e00ccdab4b3bcb9f66b512babda69ce310f217029f01a669c612c995f9e1f31R29)
* [Benchmark](https://github.com/thanos-io/thanos/pull/5474/files#diff-5e00ccdab4b3bcb9f66b512babda69ce310f217029f01a669c612c995f9e1f31R83)

No matter how hard I tried, I could not reproduce ANY race the [fix](https://github.com/minio/minio-go/pull/1357) was made for. I even try to trigger retries, and all works fine for Linux `*os.File`. I tried to look for any repro tests on your side but none was provided to prove that there is indeed any race. Perhaps the race was detected on certain OS? (windows/MacOS?)

The benchmarks show significant allocation problems while copying large files (PartSize 64MB). I also optimize naive chunk preallocations.

```
benchstat v1.txt v2-upgr.txt
name       old time/op    new time/op    delta
Upload-12     300ms ± 7%     283ms ± 7%     ~     (p=0.310 n=5+5)

name       old alloc/op   new alloc/op   delta
Upload-12    69.3MB ± 0%     0.5MB ± 3%  -99.23%  (p=0.008 n=5+5)

name       old allocs/op  new allocs/op  delta
Upload-12     3.33k ± 2%     3.16k ± 1%   -5.01%  (p=0.008 n=5+5)
```

I know you reduced default PartSize to 16MB (from 128MB) to reduce default allocation spread, but this is still unacceptable and 16MB part size have other tradeoffs (latency). So in this PR I propose to revert that fix and optimize allocations. I would love to be e able to use minio-go in Thanos and related projects, but we need this path to be leaner. One of the components Thanos sidecar is usually deployed with limited mem (100-200 MB), so optimizations in this library matters.

Perhaps if you don't want to accept this PR because you are afraid of (IMO unproven) potential races, you would be happy with adding extra option to `Put` API like `ReuseReader`? 

I am happy to be proven wrong that we indeed have raced here (e.g I did not test other OS than Linux), but we can't accept 60x allocation regression for unexisting race. Let me know what do you think, thanks! 🤗 
 
This version is also proposed to be used in Thanos: https://github.com/thanos-io/thanos/pull/5474

Fixes: https://github.com/thanos-io/thanos/issues/3967
Fixes: https://github.com/thanos-io/thanos/issues/5393

### CODE REVIEW

File: api-put-object-streaming.go [hunk 1]
```diff
@@ -229,12 +219,12 @@ func (c *Client) putObjectMultipartStreamFromReadAt(ctx context.Context, bucketN
 			return UploadInfo{}, ctx.Err()
 		case uploadRes := <-uploadedPartsCh:
 			if uploadRes.Error != nil {
+
 				return UploadInfo{}, uploadRes.Error
 			}
 
 			// Update the totalUploadedSize.
 			totalUploadedSize += uploadRes.Size
-			// Store the parts to be completed in order.
```
Comment (harshavardhana) [inline api-put-object-streaming.go:237]: Keep the comments @bwplotka 

```
// Store the parts to be completed in upload order.
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
Let me think about this more.

File: api-put-object-streaming.go (hunk 1 above)
Comment (bwplotka) [reply api-put-object-streaming.go:237]: Yea, this comment was quite confusing we don't store things in order, we only sort it later. Happy to make it clear in this comment. 🤗

**Comment by klauspost** [comment]
@bwplotka 

`ReadAt` is concurrency safe on `*os.File`, but the `io.ReaderAt` interface does *not* mandate that this is safe for all implementations.

Therefore assuming that it is, can lead to subtle bugs.

**Comment by bwplotka** [comment]
Ack.

Would you accept an explicit option to reuse reader in options then? (:

**Comment by harshavardhana** [comment]
> Ack.
> 
> Would you accept an explicit option to reuse reader in options then? (:

what I would suggest is we split the function instead of a generic ReaderAt make that ReaderAt by explicit *os.File - rest of them pay the penalty of being io.Reader instead since we can't reliably guarantee that concurrency is possible.

**Review by harshavardhana — CHANGES_REQUESTED** [review]
```go
func (c *Client) putObjectMultipartStreamFromReadAt(ctx context.Context, bucketName, objectName string,
	reader *os.File, size int64, opts PutObjectOptions,
```

Just this change would provide the necessary guarantees

**Comment by harshavardhana** [comment]
The actual issue because of a "retry" - I reproduced the problem even with this PR

```
==================
WARNING: DATA RACE
Write at 0x00c0034d2a08 by goroutine 132:
  io.(*SectionReader).Seek()
      /home/harsha/sdk/go1.18.3/src/io/io.go:534 +0x165
  github.com/minio/minio-go/v7.Client.executeMethod()
      /home/harsha/go/src/github.com/minio/minio-go/api.go:564 +0x4dd
  github.com/minio/minio-go/v7.Client.uploadPart()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object-multipart.go:282 +0x533
  github.com/minio/minio-go/v7.Client.putObjectMultipartStreamFromReadAt.func2()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object-streaming.go:171 +0x651
  github.com/minio/minio-go/v7.Client.putObjectMultipartStreamFromReadAt.func3()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object-streaming.go:192 +0x47

Previous write at 0x00c0034d2a08 by goroutine 401:
  ??()
      -:0 +0xffffffffffffffff
  net.(*conn).Write()
      /home/harsha/sdk/go1.18.3/src/net/net.go:195 +0xa4
  net.(*TCPConn).Write()
      <autogenerated>:1 +0x64
  net.(*writerOnly).Write()
      <autogenerated>:1 +0x76
  io.copyBuffer()
      /home/harsha/sdk/go1.18.3/src/io/io.go:428 +0x2dc
  io.Copy()
      /home/harsha/sdk/go1.18.3/src/io/io.go:385 +0x8d
  net.genericReadFrom()
      /home/harsha/sdk/go1.18.3/src/net/net.go:662 +0x28
  net.(*TCPConn).readFrom()
      /home/harsha/sdk/go1.18.3/src/net/tcpsock_posix.go:54 +0xcc
  net.(*TCPConn).ReadFrom()
      /home/harsha/sdk/go1.18.3/src/net/tcpsock.go:130 +0x68
  io.copyBuffer()
      /home/harsha/sdk/go1.18.3/src/io/io.go:412 +0x1c2
  io.Copy()
      /home/harsha/sdk/go1.18.3/src/io/io.go:385 +0x86
  net/http.persistConnWriter.ReadFrom()
      /home/harsha/sdk/go1.18.3/src/net/http/transport.go:1775 +0x18
  bufio.(*Writer).ReadFrom()
      /home/harsha/sdk/go1.18.3/src/bufio/bufio.go:768 +0x28d
  io.copyBuffer()
      /home/harsha/sdk/go1.18.3/src/io/io.go:412 +0x1c2
  io.Copy()
      /home/harsha/sdk/go1.18.3/src/io/io.go:385 +0x64
  net/http.(*transferWriter).doBodyCopy()
      /home/harsha/sdk/go1.18.3/src/net/http/transfer.go:411 +0x3e
  net/http.(*transferWriter).writeBody()
      /home/harsha/sdk/go1.18.3/src/net/http/transfer.go:369 +0x799
  net/http.(*Request).write()
      /home/harsha/sdk/go1.18.3/src/net/http/request.go:698 +0x1197
  net/http.(*persistConn).writeLoop()
      /home/harsha/sdk/go1.18.3/src/net/http/transport.go:2395 +0x2e4
  net/http.(*Transport).dialConn.func6()
      /home/harsha/sdk/go1.18.3/src/net/http/transport.go:1751 +0x39

Goroutine 132 (running) created at:
  github.com/minio/minio-go/v7.Client.putObjectMultipartStreamFromReadAt()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object-streaming.go:151 +0x79d
  github.com/minio/minio-go/v7.Client.putObjectMultipartStream()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object-streaming.go:49 +0x31a
  github.com/minio/minio-go/v7.Client.putObjectCommon()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object.go:241 +0x9f7
  github.com/minio/minio-go/v7.Client.PutObject()
      /home/harsha/go/src/github.com/minio/minio-go/api-put-object.go:208 +0x384
  main.main.func1()
      /home/harsha/go/src/github.com/minio/minio-go/race-test.go:95 +0x493
  main.main.func3()
      /home/harsha/go/src/github.com/minio/minio-go/race-test.go:111 +0x47
```

You are not able to reproduce this because you are not seeing the issue with retrying that is 
triggered when one of the parallel uploads fails - since there is a Seek() call to re-upload, that
is inherently racy @bwplotka 

To avoid that Seek we have to make sure that Seeking is not allowed if you are going to
use ReaderAt without buffering.

Instead, the Seek should be converted into a ReaderAt call again to avoid offset in parallel
or add a locked "SectionReader"

I reproduced it by taking the server up and down multiple times - this triggers the race.

**Comment by bwplotka** [comment]
Ack. I manually injected fault to "executeMethod" to trigger the race, but unsuccessfully. 

Thanks for repro, it seems we know what we are looking for now then!

**Comment by harshavardhana** [comment]
Let me fix it in your PR and send a PR to your fork - we can then merge it.

**Comment by bwplotka** [comment]
Amazing 😱 

Your are the best all! <#

**Pull request merged**